kuksa_databroker/databroker/src/grpc/kuksa_val_v1/val.rs

```diff
@@ -375,31 +375,55 @@ impl proto::val_server::Val for broker::DataBroker {
             ));
         }
 
-        let mut entries = HashMap::new();
-
-        for entry in request.entries {
-            let mut fields = HashSet::new();
-            for id in entry.fields {
-                match proto::Field::from_i32(id) {
-                    Some(field) => match field {
-                        proto::Field::Value => {
-                            fields.insert(broker::Field::Datapoint);
-                        }
-                        proto::Field::ActuatorTarget => {
-                            fields.insert(broker::Field::ActuatorTarget);
+        let mut valid_requests: HashMap<String, (regex::Regex, HashSet<broker::Field>)> =
+            HashMap::new();
+
+        for entry in &request.entries {
+            if entry.path.contains('*') && !glob::is_valid_pattern(&entry.path) {
+                tonic::Status::new(tonic::Code::InvalidArgument, "Invalid Pattern Argument");
+                continue;
+            }
+
+            let regex_exp = glob::to_regex(&entry.path);
+            if let Ok(regex) = regex_exp {
+                let mut fields = HashSet::new();
+                for id in &entry.fields {
+                    if let Some(field) = proto::Field::from_i32(*id) {
+                        match field {
+                            proto::Field::Value => {
+                                fields.insert(broker::Field::Datapoint);
+                            }
+                            proto::Field::ActuatorTarget => {
+                                fields.insert(broker::Field::ActuatorTarget);
+                            }
+                            _ => {
+                                // Just ignore other fields for now
+                            }
                         }
-                        _ => {
-                            // Just ignore other fields for now
+                    };
+                }
+                valid_requests.insert(entry.path.clone(), (regex, fields));
+            }
+        }
+
+        let mut entries: HashMap<i32, HashSet<broker::Field>> = HashMap::new();
+
+        if !valid_requests.is_empty() {
+            for (_, (regex, fields_set)) in valid_requests {
+                broker
+                    .for_each_entry(|entry| {
+                        let entry_path = &entry.metadata().path;
+                        if regex.is_match(entry_path) {
+                            entries
```

Comment:
> Looks better.
> 
> It still doesn't handle the case when trying to subscribe to a non existing path.
> Since everything is treated as a wildcard pattern, I guess it would have to be solved by checking if "wildcard pattern matches nothing" -> error "not found".

Something along these lines.
```diff
+           let mut requested_path_found = false;
            for (_, (regex, fields_set)) in valid_requests {
                broker
                    .for_each_entry(|entry| {
                        let entry_path = &entry.metadata().path;
                        if regex.is_match(entry_path) {
+                           requested_path_found = true;
                            entries
            ...
+           if !requested_path_found {
+               // Return error "Not found"
+           }
```

Reply:
it should be include inside the for loop I suppose, right? We want to return an error in case any path dont have any entry